### PR TITLE
exercise01: Feature/rest api for songs

### PR DESCRIPTION
**Pull request by sayildiz**
### Configurations and Deployment
- Heroku Deployment: https://kbe-songify.herokuapp.com/
- API Endpoint : https://kbe-songify.herokuapp.com/Songify
- Swagger Ui with SpringFox : https://kbe-songify.herokuapp.com/swagger-ui.html#/song-controller

### Example Postman Tests
- duplicate Songs by Title and Artist returns HTTP Code 409 Conflict and body contains string text with "already exists at Songify/songs/" + id"

- getSongById with non existing id returns HTTP Code 404 Not Found
curl --location --request GET 'https://kbe-songify.herokuapp.com/Songify/songs/345'

- non Existing api endpoint returns Http Code 500 Bad Request by default by Spring
curl --location --request GET 'https://kbe-songify.herokuapp.com/Songify/songs/einvalsadkjö'

- deleting Songs twice returns HTTP Code 204 (always returns 204)

### CI/CD
created CI/CD pipeline for this branch only for testing, should be applied later on to only main branch or develop

1. test with maven
2. build with maven and deploy on heroku
3. test api with exported Postman Collection using node/newman

### Notes
After build and API-test on Heroku app which inserts a Song and deletes it the DB is different to initial state.
Duplicate run after build of Postman Collection will fail.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  build:
+    runs-on: ubuntu-latest
+    needs: test
+    steps:
+      - uses: actions/checkout@v2
+      - uses: akhileshns/heroku-deploy@v3.12.12 # This is the action
+        with:
+          heroku_api_key: ${{secrets.HEROKU_API_KEY}}
+          heroku_app_name: "kbe-songify" #Must be unique in Heroku
+          heroku_email: "y.samet97@gmail.com"
```
Comment (htwb-steven) [inline .github/workflows/main.yml]: I recommend not including private emails in your repository.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
+name: Test, Build, Deploy and Api-Test
+
+on:
+  push:
+    branches:
```
Comment (htwb-steven) [inline .github/workflows/main.yml:5]: You can probably use this pipeline for the other exercises as well

File: Songify/postman_collection.json [hunk 3]
```diff
@@ -0,0 +1,653 @@
+{
```
Comment (htwb-steven) [inline Songify/postman_collection.json:1]: Missing test:
- POST with invalid JSON in payload e.g. something is missing or wrongly spelled

**Review by htwb-steven — APPROVED** [review]
LGTM. Good solution!

File: Songify/postman_collection.json (hunk 3 above)
Comment (sayildiz) [reply Songify/postman_collection.json:1]: added 4 new tests with 
- invalid JSON Payload
- missing value
- missing json member and value
- wrongly spelled json member

Example wrongly spelled: 

`curl --location --request POST 'https://kbe-songify.herokuapp.com/Songify/songs' \
--header 'Content-Type: application/json' \
--data-raw '{
    "titl": "Wrecking Ball",
    "artist": "MILEY CYRUS",
    "label": "RCA",
    "relased": 2013
}'`

File: .github/workflows/main.yml (hunk 1 above)
Comment (sayildiz) [reply .github/workflows/main.yml]: removed email with Github Secrets even though you can see it in my git history :)

Comment (htwb-steven) [reply .github/workflows/main.yml]: A solution for this is to delete PR + branch and rewrite git history

**Pull request merged**